File: KelpNet/Optimizers/Adam.cs
﻿using System;
using System.Collections.Generic;
using KelpNet.Functions;

namespace KelpNet.Optimizers
{
    public class Adam : Optimizer
    {
        private double alpha;
        private double beta1;
        private double beta2;
        private double eps;

        double lr
        {
            get
            {
                double fix1 = 1 - Math.Pow(this.beta1, this.t);
                double fix2 = 1 - Math.Pow(this.beta2, this.t);
                return this.alpha * Math.Sqrt(fix2) / fix1;
            }
        }

        private NdArray[][] m;
        private NdArray[][] v;

        public Adam(double alpha = 0.001, double beta1 = 0.9, double beta2 = 0.999, double eps = 1e-8)
        {
            this.alpha = alpha;
            this.beta1 = beta1;
            this.beta2 = beta2;
            this.eps = eps;
        }

        protected override void DoUpdate(List<OptimizableFunction> optimizableFunctions)
        {
            for (int i = 0; i < optimizableFunctions.Count; i++)
            {
                for (int j = 0; j < optimizableFunctions[i].Parameters.Count; j++)
                {
                    for (int k = 0; k < optimizableFunctions[i].Parameters[j].Length; k++)
                    {
                        double grad = optimizableFunctions[i].Parameters[j].Grad.Data[k];

                        m[i][j].Data[k] += (1 - this.beta1) * (grad - m[i][j].Data[k]);
                        v[i][j].Data[k] += (1 - this.beta2) * (grad * grad - v[i][j].Data[k]);

                        optimizableFunctions[i].Parameters[j].Param.Data[k] -= lr * m[i][j].Data[k] / (Math.Sqrt(v[i][j].Data[k]) + this.eps);
                    }
                }
            }
        }

        public override void Initialize(FunctionStack fs)
        {
            this.m = new NdArray[fs.OptimizableFunctions.Count][];
            this.v = new NdArray[fs.OptimizableFunctions.Count][];

            for (int i = 0; i < fs.OptimizableFunctions.Count; i++)
            {
                this.m[i] = new NdArray[fs.OptimizableFunctions[i].Parameters.Count];
                this.v[i] = new NdArray[fs.OptimizableFunctions[i].Parameters.Count];

                for (int j = 0; j < fs.OptimizableFunctions[i].Parameters.Count; j++)
                {
                    this.m[i][j] = NdArray.ZerosLike(fs.OptimizableFunctions[i].Parameters[j].Param);
                    this.v[i][j] = NdArray.ZerosLike(fs.OptimizableFunctions[i].Parameters[j].Param);
                }
            }
        }
    }
}

File: src/jp/nyatla/kelpjava/optimizers/Adam.java
﻿package jp.nyatla.kelpjava.optimizers;

import jp.nyatla.kelpjava.OptimizerParameter;
import jp.nyatla.kelpjava.functions.common.FunctionParameter;
import jp.nyatla.kelpjava.optimizers.common.Optimizer;

/**
 * [Serializable]
 */
public class Adam extends Optimizer {
	private static final long serialVersionUID = 2087321425467232653L;
	final public double Alpha;
	final public double Beta1;
	final public double Beta2;
	final public double Epsilon;

	private long UpdateCount = 1;
	/**
	 * コピーコンストラクタ
	 * @param i_src
	 */
	private Adam(Adam i_src) {
		super(i_src);
		this.Alpha = i_src.Alpha;
		this.Beta1 = i_src.Beta1;
		this.Beta2 = i_src.Beta2;
		this.Epsilon = i_src.Epsilon;
	}

	public Adam() {
		this(0.001);
	}
	public Adam(double i_alpha) {
		this(i_alpha, 0.9, 0.999, 1e-8);
	}

	public Adam(double i_alpha, double i_beta1, double i_beta2, double i_epsilon) {
		super();
		this.Alpha = i_alpha;
		this.Beta1 = i_beta1;
		this.Beta2 = i_beta2;
		this.Epsilon = i_epsilon;
	}

	public void addFunctionParameters(FunctionParameter[] i_functionParameters) {
		this.optimizerParameters = new OptimizerParameter[i_functionParameters.length];

		for (int i = 0; i < this.optimizerParameters.length; i++) {
			this.optimizerParameters[i] = new AdamParameter(
					i_functionParameters[i], this);
		}
	}
	@Override
	public Optimizer deepCopy() {
		return new Adam(this);
	}
	private class AdamParameter extends OptimizerParameter {
		private static final long serialVersionUID = -2950834706490917841L;

		final private Adam optimiser;

		private final double[] m;
		private final double[] v;

		public AdamParameter(AdamParameter i_src) {
			super(i_src);
			this.optimiser = (Adam) i_src.optimiser.deepCopy();
			this.m = i_src.m.clone();
			this.v = i_src.v.clone();
		}

		public AdamParameter(FunctionParameter parameter, Adam optimiser) {
			super(parameter);
			this.m = new double[parameter.length()];
			this.v = new double[parameter.length()];
			this.optimiser = optimiser;
		}

		@Override
		public void updateFunctionParameters() {
			double fix1 = 1 - Math.pow(this.optimiser.Beta1,this.optimiser.UpdateCount);
			double fix2 = 1 - Math.pow(this.optimiser.Beta2,this.optimiser.UpdateCount);
			double lr = this.optimiser.Alpha * Math.sqrt(fix2) / fix1;

			for (int i = 0; i < this.functionParameters.length(); i++) {
				double grad = this.functionParameters.grad.data[i];

				this.m[i] += (1 - this.optimiser.Beta1) * (grad - this.m[i]);
				this.v[i] += (1 - this.optimiser.Beta2)
						* (grad * grad - this.v[i]);

				this.functionParameters.param.data[i] -= lr * this.m[i]
						/ (Math.sqrt(this.v[i]) + this.optimiser.Epsilon);
			}
			return;
		}

		@Override
		public Object deepCopy() {
			return new AdamParameter(this);
		}
	}


}
